src/views/Quest.vue

```diff
@@ -0,0 +1,304 @@
+import ChatManager from '../ChatManager';
+
+
+export default Vue.extend({
+    data() {
+        return {
+            isLoading: true,
+            chat: <ChatManager | null>null,
+            logoSourcePng: require('../assets/logo_eterna.svg'),
+        };
+    },
+    async mounted() {
+        try {
+            await this.$store.dispatch(Action.GET_QUEST_ACHIEVEMENT_ROADMAP);
+            await this.$store.dispatch(Action.GET_PUZZLES, new URLSearchParams({type: 'puzzles', ...this.$route.query}));
+            this.chat = new ChatManager('chat-container', this.$store);
+            this.isLoading = false;
+        } catch (error) {
+            console.error(error);
+        }
+    },
+    components: {
+        Carousel,
+        HeaderBar,
+        NavBar,
+        ProgressBar,
+        TutorialCard,
+        PuzzleCard
+    },
+    computed: {
+        achievement(): Achievement {
+            return (this.$store.state.quest_roadmap as Achievement[]).find(a => a.key == this.$route.params.id && a.level == +this.$route.params.level)!;
+        },
+        puzzles(): (PuzzleItem & {cleared: boolean})[] {
+            const puzzleList = this.$store.state.puzzle_list as PuzzleList;
+            const puzzlesWithCleared = puzzleList.puzzles.map(puzzle => ({
+                ...puzzle,
+                cleared: puzzleList.cleared.some(cleared => cleared.nid === puzzle.id)
+            }));
+
+            // Sort such that puzzle A which specifies its next puzzle is puzzle B is sorted before puzzle A
+            // The first puzzle is the one that has no other puzzle pointing to it
+            const orderedPuzzles: (PuzzleItem & {cleared: boolean})[] = [];
+            let puzzle = puzzlesWithCleared.find(
+                candidatePuzzle => !puzzlesWithCleared.some(otherPuzzle => otherPuzzle['next-puzzle'] === candidatePuzzle.id)
+            );
+            while (puzzle) {
+                orderedPuzzles.push(puzzle);
+                const nextPuzzle = puzzle['next-puzzle'];
+                puzzle = puzzlesWithCleared.find(candidatePuzzle => candidatePuzzle.id === nextPuzzle);
+            }
+            // Add any additional puzzles not part of the next puzzle "chain"
+            orderedPuzzles.push(...puzzlesWithCleared.filter(candidatePuzzle => !orderedPuzzles.includes(candidatePuzzle)));
+            return orderedPuzzles;
+        },
+    },
+    methods: {
+        openChat() {
+            if (this.chat) {
+                this.chat.toggleVisibility();
+            }
+        },
+        resolveUrl(path: string) {
+            if (path.startsWith('http')) return path;
+            if (path.startsWith('/')) return process.env.APP_SERVER_URL + path;
+            return process.env.APP_SERVER_URL + '/' + path;
+        },
+        getPuzImg(nid: string | null){
+            return (
+            nid &&
+            `https://renderv2-prod-renderv2bucket86ab868d-1aq5x6e32xf92.s3.amazonaws.com/puzzle_mid_thumbnails/thumbnail${nid}.svg`
```

Comment:
Hard-coding the server here seems brittle. Is the server-rendering thumbnails something we'll address in the rewrite?

Reply:
I think we decided that the full thumbnail URL will be stored in the rewrite